### ISSUE TITLE
Unsolved distant problems

### ISSUE THREAD

**Issue by GrapesGoober** [GrapesGoober/simplic-py#22]
# These are problems you're not going to face any time soon
As the name suggests. This will be a place for you to comment some design or thought issues that is not yet resolved.

**Comment by GrapesGoober** [comment]
# Handling stack overflow
The simplest design for the overflow trigger is check for availibility bitmap. As soon as the call stack overlaps the allocated heap (i.e. when they collides) then throw the stack overflow exception.
# Problem
The function call is handled by the IC translator, but the heap allocation is going to be built as subroutines (with global variables for bitmaps). This means that for the function call to detect this, the IC translator has to be able to read the bitmap as well.
# Solutions?
### Built-in heap allocation
Have the heap allocation be built-in as a feature of the IC? I.e. much like call-overhead, have the heap allocation be handled behind the scenes. This also makes the GC be built-in as well... not good
### Built-in superglobal bitmap variables for the IC
The IC will handle the bitmap, but the allocation will still be done as subroutines. This also allow the GC to be subroutines. As the IC will be handling the bitmaps, the IC will provide syntax to read/modify those bitmaps. It'll also do the checks as well

**Comment by GrapesGoober** [comment]
# How shall the GC works?
## Idea 1 - Free all
- Since the function knows all the metadata for all heap variables, the function itself will GC
- First: free all heap variables
- Second: start from main (bottom of call stack), then iterates over each heap item to re-allocates
### Problem
- Large time & space complexity
- The function itself needs to know which variables it needs to de-allocates. When looking at the call-stack alone, you don't have that info. A function might me incomplete
```
c = a + b
d = c * a
e = dothing(c, d)
f = e * 2
```
- In this example, looking the call stack alone will not see the "e" and "f" variable, since those aren't finished executing yet.
### Solution
Store every possible call-points of a function, and generate GC subroutine for all of those possibilities

### Idea 2 - Selective free
- The function identifies ALL the heap variables it has allocated
- It frees selectively using its own entry points
- Then it uses GC to re-alloc
- Would trigger if the function has freed something (i.e. has overwrite some pointer, potentially leaving something unreachable)

**Comment by GrapesGoober** [comment]
# GC sweeping runtime-determined sizes
This problem primarily concerns arrays. The gist is that for struct types, such as lists (where each item is implemented as nodes) or dict (again, as nodes), those have compile-time determined size. The problem occurs on arrays, where its size isn't known at compile time. This makes the GC much harder to code to handle sweeping.
## Solution 1 - Runtime "size" variable
This solution proposes that all heap data begins with a variable "size", even for struct types. This simplifies GC such that it only needs to look for this variable to perform the sweep. Note that this could also have a variant where only the array types have this variable
### Problems
- This doesn't actually simplifies the GC's compile-time generation. It still needs the compile-time symbol table to determine if/where the object has any nested pointers. In fact, putting this variable into runtime might actually complexify the GC, as it _might_ need to look into runtime variables as well (although it already is doing with pointer traversals anyways)

## Solution 2 - Built-in
Built-in data structures are handled by the compiler itself, i.e. the logic for traversal & GC generation is handled by the IR. In other words, the high-level syntax is not powerful enough to handle these things, so the task is managed to the IR instead. 

One idea is that ALL built-ins are implemented as IR, and GC code is also manually written. This covers arrays, strings, lists, dicts, and graphs. Zero-cost abstractions can actually make this more appealing to use, and the programmer can still write complex structures (with reference types, such as graphs).

The makes the built-ins less transparent, and requires a system API to interact between high-level and IR (say, high-level wants to do certain things such as iterate over arrays. Since implementation over arrays are handled by IR instead, there needs to be some sort of interface for high-level abstractions to interact directly with the IR that the programmer wants to iterate).

**Comment by GrapesGoober** [comment]
# COOL IDEA - LITTER FUNCTION DETECTION
This is an internal compiler thing, and it's done automatically behind the scenes for you. The idea is that the compiler will, as it compiles a function, it will determine if that function is a litter function or not. A litter function is essentially a function that might produce garbage (hence the name) and thus necessitate a need for GC trigger. A non-litter function is a function that garantees to not produce garbage, even though it can still mutate heap variables somewhat.
### Definition
A litter function is any function that does any of the following
- mutates a pointer of an argument. For example, inserting to a list. Mutating non-pointers are not considered littering.
- mutates a pointer of a return. For example, creating a list and returning said list. The act of creating list requires multiple pointer mutation.

This check can be done in the compile time, or it can be done at runtime. To check at runtime, maybe have a single header variable somewhere (a boolean) that denotes whether this particular function has littered or not. Upon return, this "litter" variable will be used to check whether to call the GC. After the GC is called, the function will be considered "clean" as it's got rid of its own litter.

**Comment by GrapesGoober** [comment]
# Memory Allocation and Garbage Collection
Using DSW algorithm, can read more on [this](https://xlinux.nist.gov/dads/HTML/SchorrWaiteGraphMarking.html), and [this](https://xlinux.nist.gov/dads/HTML/binaryTreeRepofTree.html)
Can potentially adapt this to simplic's availability bitmap and struct's pointer handling

# Partitioning
As for memory system, might be considering using partitions, where the entire memory is divided in halves. One half for heap, one half for call stack and GC stack. If using DSW then this is not needed, as a simple heap -> <- stack structure should work just fine

**Comment by GrapesGoober** [comment]
# COOL IDEA - LITTER FUNCTION DETECTION
This is an internal compiler thing, and it's done automatically behind the scenes for you. The idea is that the compiler will, as it compiles a function, it will determine if that function is a litter function or not. A litter function is essentially a function that might produce garbage (hence the name) and thus necessitate a need for GC trigger. A non-litter function is a function that garantees to not produce garbage, even though it can still mutate heap variables somewhat.
## Definition
A litter function is any function that does any of the following
- mutates a pointer of an argument. For example, inserting to a list. Mutating non-pointers are not considered littering.
- mutates a pointer of a return. For example, creating a list and returning said list. The act of creating list requires multiple pointer mutation.

## Runtime Littering Check
This littering check can be done either at compile time or at runtime. To check at runtime, maybe have a single function-header variable somewhere (a boolean) that denotes whether this particular function has littered or not, i.e. a 'isLitter' variable. Upon return, this 'isLitter' variable will be used to check whether to call the GC. After the GC is called, the function will be considered "clean" as it's got rid of its own litter. The caller does not need to worry about the garbage this function produces

## Littering Bitmap
One possible way to store 'isLitter' variable is to use bitmap. Instead of mapping whether the function litter as a whole (which will force the GC to check for everything, the arguments, the returns, etc), we use a bitmap to denote which part of the function does it litter. Say, the index zero denotes that the "return" value litters, and the rest denotes which argument litters. In this case, the GC will know which entry point to check for, as it has a clear root of the traversal. This can even be benificial in runtime case, where the GC can narrow down which variable to check for.
### Potential for special instruction?
Could there be a dedicated `bitset` or `bittest` specifically for bitmaps? This could complement `clz` quite nicely, however the `mode2` bits encourage compression

## Litter array
Instead of having a fixed set of potential entry points, perhaps we allocate a global array of some size, to be used as "litter entry points". In this case, upon littering, the litterer will place the point of littering onto the array. Determining which constitutes as a litter point can be complex, as not all pointer mutation cause garbage (say, internal lists can get inserted, but those will not be considered litter points as it's only internal)
### Problems
- Determining litter points, as it can be complex which subset of all pointer mutations are actually  littering.
- Fixed sized array may limit the number of littering points. In which case, the GC might need to resort to litter bitmap. This will make the litter array a performance optimization instead of a required feature, which isn't good

**Comment by GrapesGoober** [comment]
# Simplic Notebook
Let's see how this would work. Ideally, the parser will only isolate the codeblocks within the markdown and parse them as if in the same file.
## Neural Network
This is just an example of running some neural networks. We might want to have the emulation running while also reading/writing files into the emulation’s input buffer.

Maybe have the dataset be small enough? Maybe a 16x16 normalized discrete images of circles and non circles. Maybe might have this train for multiple sessions emulation sessions.

```
import NeuralNetwork, Dataset
func main() {
    let layers = createLayers() 
    let model = NeuralNetwork.Sequential
    model.train(Dataset.circles, testRatio=0.2, epoch=100)
}
```

## Multiple Cells
It is not required to write cells sequentially with its order of execution. It will still be executed on the main function regardless, and the parser will generate the symbols appropriately as it parses. Imports can also be dispersed for each cells.
```
func createLayers() {
    let layers = [
        Conv( n=16, kernel=[2, 2], activation=relu ),
        MaxPool(2, 2),
        Conv( n=16, kernel=[2, 2], activation=relu ),
        MaxPool(2, 2),
        Dense( n=128, activation=sigmoid )
    ]
}

```

**Comment by GrapesGoober** [comment]
# Simplic Syntax
When parsing, the compiler will isolate only the code cells. Can include code cells, but the order of execution does not matter; all codes will still have one entry point at function main.
```
func main () {
    let x = 3, y = 2
    print(f"the sum of numbers = {x + y}")
}
```
## Objective
This file is to demonstrate how the syntax of simplic should look like. Specifically, I'm conflicted with the `: type` similar to typescript, golang, or some other newer languages. This should make the code much easier to parse. Simplic is statically typed, but also use implicit typing in many cases. Regardless, the syntax for type specification has to be consistent for variables, arguments, and function return types. 
```
func helperFunctions : float (meshPoints : array<int>) {
    let sum = 0
    for i in meshPoints {
        # can optionally specify type explicitly
        let v : float = fixResult(i)
        sum += v / meshPoints.length
    }
    return sum
}
```

## Generic Structs and Functions
```
struct Entity<T> {
    let x : T
    let y : int
}

func CollectEntity<T> : Array<Entity<T>> (arr : Array<T>) {
    let l = Array<Entity<T>>{ size: arr.length }
    for i, a in enumerate(arr) {
        l[i] = Entity<T>{ x = a, y = i }
    }
    return l
}

```

**Comment by GrapesGoober** [comment]
# Handling stack frame
A single call stack can have many variables, but 16 immediate addressing is nit enough. As such, there should be something like this:
- P represent stack pointer
- FP represent function header pointer

FP will always point to the base of the function (the header of the function) whereas P is kept moving like a window. This allows the a "sliding window" to move around the function frame, whereas the FP still points to the function header.

Note that, with function header info, we have function ID and back pointer, so we also know the entire trace of the call stack, along with their argument values.

**Comment by GrapesGoober** [comment]
# Handling Generics in IR
Just got a really cool idea of how generic system could be handled in IR code. Here's the _why_ and the _how_:
## The Why

## The How
Use dynamic jumping! The idea is that each struct would have their own ways of doing things, so you can encapsulate struct operations as functions. This include getter, setter, and really anything you might want to do to a struct type. Since they are functions, they could be called. To achieve generic-like behaviour, use dynamic jumping! Simply put the function address there and the generic algorithm will handle it for you!

For example, let's have a generic list traversal, where each node can hold anything. Let's assume we want to traverse the current node's data first before iterating to the next node. You can write your generic traversal as such
```
function generic_traversal arg root g_func
```
the `root` argument is the start of the traversal, and the `g_func` argument is the address of the desired traversal function for the data itself. Note that the data could be anything, so it's up to the caller to define what data representation (and thus `g_func`) should look like.
```
function generic_traversal arg root g_func
label start
  # in this example, a node is s-list
  # first cell is data, second cell is next pointer
  # so first thing to do is to call generic function
  setarg root       # performing a generic function call for the data
  storem 0 g_func   # realistically, might have the IR/ASM resolve label's address instead
  add next root 1   # location address of the 'next' pointer
  setarg next 
  call generic_traversal 
  return
```

**Comment by GrapesGoober** [comment]
# Simplic Syntax 2
```
def main () {
    let x = 3, y = 2
    print(f"the sum of numbers = {x + y}")
}
```
```
def helperFunctions (array<int> meshPoints) -> float {
    let sum = 0
    for i in meshPoints {
        # can optionally specify type explicitly
        let float v = fixResult(i)
        sum += v / meshPoints.length
    }
    return sum
}
```

## Generic Structs and Functions
```
struct Entity<T> {
    T x
    int y
}

def CollectEntity<T> (Array<T> arr) -> Array<Entity<T>>{
    let l = Array<Entity<T>>{ size: arr.length }
    for i, a in enumerate(arr) {
        l[i] = Entity<T>{ x = a, y = i }
    }
    return l
}

```

**Comment by GrapesGoober** [comment]
# Handle Generics At Runtime?
```
struct Entity {
	any x
}
```

They `any` type is essentially a dynamic typing system, where it uses 2 field: `type` field (id header) and `data` field (pointer or value). Handling generics looks as follows:

```
def createRecords(any data) -> List {
	let l = []
	for i in range(0, 10) {
		l.append(randomize(data))
	}
	return l
}
```
Note that this doesn't feel like a generic system, more like dynamic typing system. This does make it simpler, but at runtime memory cost, performance cost, and potential runtime errors (since type safety is not determined at compile time)... Can we mitigate this using interface?
```
def printEachItem(List:iter l)
```

**Comment by GrapesGoober** [comment]
# Error Handling Syntax
Thinking of how should errors be handled? I want simplicity of Golang's explicit `if err {}` with the convenience of exceptions. The function `do_risky` here might throw errors, let's see how this could be handled.
## Standalone Try
Using a `try` keyword! Without this keyword, exceptions will not be handled (completely ignored). This is programmer-error prone since you might forget to include the `try` keyword and now your exceptions are unhandled (they just go silent). Still, a try keyword is interesting
```
def main() {
   try let x = do_risky()
}
```
## Explicit Return
What about having explicit error message returns? This could be a little confusing at first, but it makes for much more concise syntax. The error handling could also be enforced by compiler, as this syntax has to check for matching multi-return.
```
def main() {
   let x, err = do_risky() 
   let y, = do_risky()        # ignore errors
}
```
## Global Error
Global error variable? This is interesting, but the problem of unhandled errors still remains. This also mean that programmers need to remember which functions produce errors and which doesn't, which could be a little complex. 
```
def main() {
   let x = do_risky()
   if err { 
      printerr(err)
   }
}
```

**Comment by GrapesGoober** [comment]
# Idea for storing comparison results
When performing comparisons, the old idea is that the results are ignored. Instead, the comparison is left on the accumulator for the next `if` instruction to use. How about we also have a feature where comparisons can be kept on variable such that it can be used by `if` instruction ever see fit?
# Solution
Allow `dest` operand for `cmp` instruction! The idea is that `cmp` is identical to subtraction! Thus, the `dest` operand for arithmatics (which in this case encompasses `cmp`) could be optional. The HIR would compile `cmp` to LIR's `sub`, everything else being same, and let LIR compile to ASM automatically.

**Comment by GrapesGoober** [comment]
# Future High Level App
One idea for this app is a economical feasibility analysis program. This takes input of each cash flow variables and other constant figures, then computes the FV and PV cash flow for each year, NPVs, BEP, and ROI. It then prints out the numbers onto `stdout`. This reminds me of the use case of those old IBM computers running punchcard fortran programs, doing some accounting number crunches.
